### ISSUE TITLE
Tracking requests are stored but none are processed

### ISSUE THREAD

**Issue by camlcase** [piwik/plugin-QueuedTracking#194]
I have QueuedTracking with Redis and using Memurai on a Windows machine. When plugin is active, I can see items in queue but none of them are processed. 

Exception from the matomo.log:
```
#0\core\Tracker\RequestSet.php(161),
#1\plugins\QueuedTracking\Queue.php(94),
#2\plugins\QueuedTracking\Queue\Processor.php(112),
#3\plugins\QueuedTracking\Commands\Process.php(86),
#4\vendor\symfony\console\Symfony\Component\Console\Command\Command.php(257),
#5\vendor\symfony\console\Symfony\Component\Console\Application.php(874),
#6\vendor\symfony\console\Symfony\Component\Console\Application.php(195),
[internal function]: Symfony\Component\Console\Application->doRun(),
#8\core\Console.php(135),
#9\core\Access.php(670)
```

Have tried processing the queue using the "Process during tracking request" and manually via `/console queuedtracking:process`.

Example output when monitoring:
```
35 (19+16) request sets left in queue. 810.84K used memory (1020.51K peak). 0 workers active.
35 (19+16) request sets left in queue. 810.84K used memory (1020.51K peak). 0 workers active.
36 (19+17) request sets left in queue. 810.84K used memory (1020.51K peak). 0 workers active.
```

Output when processing manually:
```
Starting to process request sets, this can take a while

This worker finished queue processing with 0req/s (0 requests in 0.11 seconds)
```

I am using these versions:
PHP: 8.0
Matomo: 4.11.0
MariaDB: 10.5.9.0
Memurai: 3.1.4
PHPRedis: 5.3.7

Output from `/console queuedtracking:test`
```
Settings that will be used:
Backend: redis
NumQueueWorkers: 2
NumRequestsToProcess: 5
ProcessDuringTrackingRequest: 1
QueueEnabled: 1

Redis backend only settings (does not apply when using MySQL backend):
Host: *******
Port: 6379
Timeout: 0
Password:
Database: 0
UseSentinelBackend: 0
SentinelMasterName: mymaster

Version / stats:
PHP version: 8.0.0
Uname: Windows NT DMZWA02 10.0 build 17763 (Windows Server 2016) AMD64
PHPRedis version: 5.3.7
Backend version: 6.2.6
Memory: array (
  'used_memory' => 914360,
  'used_memory_human' => '892.93K',
  'used_memory_rss' => 856472,
  'used_memory_rss_human' => '836.40K',
  'used_memory_peak' => 1045000,
  'used_memory_peak_human' => '1020.51K',
  'used_memory_peak_perc' => '87.50%',
  'used_memory_overhead' => 850600,
  'used_memory_startup' => 685912,
  'used_memory_dataset' => 63760,
  'used_memory_dataset_perc' => '27.91%',
  'allocator_allocated' => 957376,
  'allocator_active' => 1486848,
  'allocator_resident' => 818584,
  'total_system_memory' => 0,
  'total_system_memory_human' => '0B',
  'used_memory_lua' => 37888,
  'used_memory_lua_human' => '37.00K',
  'used_memory_scripts' => 416,
  'used_memory_scripts_human' => '416B',
  'number_of_cached_scripts' => 2,
  'maxmemory' => 0,
  'maxmemory_human' => '0B',
  'maxmemory_policy' => 'noeviction',
  'allocator_frag_ratio' => 1.55,
  'allocator_frag_bytes' => 529472,
  'allocator_rss_ratio' => 0.55,
  'allocator_rss_bytes' => -668264,
  'rss_overhead_ratio' => 1.05,
  'rss_overhead_bytes' => 37888,
  'mem_fragmentation_ratio' => 1.0,
  'mem_fragmentation_bytes' => 0,
  'mem_not_counted_for_evict' => 0,
  'mem_replication_backlog' => 0,
  'mem_clients_slaves' => 0,
  'mem_clients_normal' => 164128,
  'mem_aof_buffer' => 0,
  'mem_allocator' => 'jemalloc-3.6.0',
  'active_defrag_running' => 0,
  'lazyfree_pending_objects' => 0,
  'lazyfreed_objects' => 0,
)
MaxMemory Eviction Policy config: noeviction
MaxMemory config: 0

Performing some tests:
Redis is connected: 1
Connection works in general
Success for method set(testKey, value)
Success for method setnx(testnxkey, value)
Success for method setex(testexkey, 5, value)
Success for method set(testKeyWithNx, value, array (0 => 'nx',))
Success for method set(testKeyWithEx, value, array ('ex' => 5,))
Initial expire seems to be set correctly
setIfNotExists works fine
expireIfKeyHasValue seems to work fine
Extending expire seems to be set correctly
expireIfKeyHasValue correctly expires only when the value is correct
Expire is still set which is correct
deleteIfKeyHasValue seems to work fine
List feature seems to work fine
```

**Comment by snake14** [comment]
Hi @camlcase . I'm sorry to hear that you're experiencing this issue. Have you tried searching the [Matomo forum](https://forum.matomo.org) as the warning suggested? I see that you tried monitoring the queue. Have you tried including the verbose option while manually triggering the queue, like `./console queuedtracking:process -vvv` , to see if there's any more information?

**Comment by camlcase** [comment]
Hi,

I did search for the issue but couldn´t find anything of help. 

Output when monitoring:
`40 (17+23) request sets left in queue. 728.95K used memory (1020.51K peak). 0 workers active.`

When processing requests manually using the verbose option:
```
Starting to process request sets, this can take a while

This worker finished queue processing with 0req/s (0 requests in 0.02 seconds)
```

This is what is written to the logfile when activating debug mode:
```
DEBUG QueuedTracking[2022-10-31 18:50:45 UTC] [68784] We are going to process the queue
DEBUG QueuedTracking[2022-10-31 18:50:45 UTC] [4ddb6] Nothing to notice => default behaviour
DEBUG QueuedTracking[2022-10-31 18:50:45 UTC] [4ddb6] End of the page.
DEBUG QueuedTracking[2022-10-31 18:50:45 UTC] [4ddb6] array (
DEBUG QueuedTracking[2022-10-31 18:50:45 UTC] [4ddb6] )
```

What about the software versions used? Upgrade Matomo to latest version 4.12.3? Try with an older version of PHPRedis  extension?
```
PHP: 8.0
Matomo: 4.11.0
MariaDB: 10.5.9.0
Memurai: 3.1.4
PHPRedis: 5.3.7
```

**Comment by snake14** [comment]
Thank you for trying that debugging. I think that your version numbers look alright. I must admit that I'm not very familiar with Memurai or PHPRedis though. I was hesitant to recommend upgrading Matomo, since we had a request format change in version 4.12.0 that caused problems for some QueuedTracking users. However, that should have been patched with 4.12.1, so you could try upgrading to 4.12.3. There is a slight chance that some of your existing requests might not be able to process, but I guess that's already the case.

**Comment by camlcase** [comment]
Thank you @snake14 for taking time answer my question. I will see if we can get an upgrade done. I will post an update here when this is done and tested.

_I found this issue which reminds me of my problem here:_ [https://github.com/matomo-org/plugin-QueuedTracking/issues/192](https://github.com/matomo-org/plugin-QueuedTracking/issues/192)

**Update:** I was able to read out some elements from the trackingqueue using a deflate stream, but the value could not be read into anything but an empty string. As you implied.

**Comment by snake14** [comment]
You're most welcome. Sounds good.

Yeah. That was the issue that immediately came to mind when you posted your issue, but then I saw that you're running Matomo 4.11.

**Comment by addison-grant** [comment]
We seem to be running into a possibly similar issue. Starting some troubleshooting/debugging.
```
/var/www/html # ./console queuedtracking:monitor
Queue is enabled
Request sets in the queue will be processed automatically after a tracking request
Up to 1 workers will be used
Processor will start once there are at least 25 request sets in the queue
1901332 (1901332) request sets left in queue. 4.80G used memory (4.80G peak). 1 workers active.
```

```
/var/www/html # ./console queuedtracking:process -vvv
Starting to process request sets, this can take a while

This worker finished queue processing with 0req/s (0 requests in 0.01 seconds)
```


I am trying to get some example data out of the queue. I am getting it out of our Redis docker container with
```
docker exec -it matomo_redis_1 redis-cli LRANGE "trackingQueueV1" 1 5 > out.csv
```
Then trying to process it in PHP:
```php
$file = fopen("out.csv","rb"); // not sure if I should use "rb" or "r"
$data = fgetcsv($file);
foreach ($data as $value) {
  echo gzuncompress($value);
}
```


When I try to run a value from the "trackingQueueV1" redis list, I am seeing this for each item: 
```
Warning: gzuncompress(): data error in php shell code on line 2

Call Stack:
  444.9201     518160   1. {main}() php shell code:0
  444.9204     509968   2. gzuncompress($data = 'x\\x9c\\xedX\\xebs\\x9aJ\\x1b\\xffW\\...') php shell **code:2**
```
Any tips appreciated. 

Another note, we are running two app servers, and I think I accidentally overloaded the redis container for one of them. It seems to have cleared the queue, and is processing requests again. I haven't done the same thing to the other container in case there's valueable debugging info on it.

From app server that is processing again:
```
[root@matomo ~]# docker exec -it matomo_app_1 ./console queuedtracking:monitor
Queue is enabled
Request sets in the queue will be processed automatically after a tracking request
Up to 1 workers will be used
Processor will start once there are at least 25 request sets in the queue
4 (4) request sets left in queue. 400.16K used memory (889.62K peak). 0 workers active.
```

I'm not 100% sure if it means this node is processing properly.

**Comment by addison-grant** [comment]
My goal above is to help see if we can identify certain structures that are messing things up. Also, I believe we're currently running Maotmo 4.12.2 off the docker image with a `redis:6.2-alpine` container. Let me know if I can provide more info.

**Comment by heurteph-ei** [comment]
@camlcase @snake14 
I see no issue in your case:
> `NumQueueWorkers: 2`
> `NumRequestsToProcess: 25`
> `36 (19+17) request sets left in queue. 810.84K used memory (1020.51K peak). 0 workers active.`

As 19 < 25 and 17 < 25, in my opinion, both queues should not be processed until they reach 25!

Also, there is something I don't understand:
> `QueueEnabled: 0`

-> Queues should not run...

**Comment by camlcase** [comment]
Hi @heurteph-ei and thank you,

You´re right. I did run the plugin test command with queue disabled. I´ve updated the original issue with new info with queue enabled. The information almost looks the same.

I´ve lowered the threshold to 5 (originally 25) and did some monitoring and it doesn´t make sense, at least not for me. The requests never reaches threshold. The request count walks up and down. No workers ever become active and no requests are processed when running the queuedtracking:process command manually.
```
6 (4+2) request sets left in queue. 852.01K used memory (1020.51K peak). 0 workers active.
6 (4+2) request sets left in queue. 852.01K used memory (1020.51K peak). 0 workers active.
6 (4+2) request sets left in queue. 872.49K used memory (1020.51K peak). 0 workers active.
6 (4+2) request sets left in queue. 872.49K used memory (1020.51K peak). 0 workers active.
2 (0+2) request sets left in queue. 872.27K used memory (1020.51K peak). 0 workers active.
4 (2+2) request sets left in queue. 872.46K used memory (1020.51K peak). 0 workers active.
4 (2+2) request sets left in queue. 872.46K used memory (1020.51K peak). 0 workers active.
4 (2+2) request sets left in queue. 872.46K used memory (1020.51K peak). 0 workers active.
```
The queuedtracking:test command creates a fooList key in Redis which value I can decompress and read. But the values in the trackingQueueV1 can´t be read for some reason.

**Comment by heurteph-ei** [comment]
As `ProcessDuringTrackingRequest: 1`, then the queue is processed by the HTTP request as soon as it reaches 5.
The 1st queue was filled with 4 requests, then reached 5 (and was processed by the 5th request), and became empty (0)...
You server seems to be not high loaded... Then the 2nd queue seems to not be used often. And gets time to be filled...

**Comment by camlcase** [comment]
@heurteph-ei Ok I see. I did some more testing.  I disabled the `ProcessDuringTrackingRequest` and did manual verbose processing with `NumRequestsToProcess: 5`.

Before running manual process:
`97 (55+42) request sets left in queue. 934.98K used memory (1.01M peak). 0 workers active.`

After running manual process:
`2 (0+2) request sets left in queue. 933.79K used memory (1.01M peak). 0 workers active.`

The result says:
`This worker finished queue processing with 0req/s (0 requests in 0.95 seconds)`

But with this WARNING for each request:
```
[2022-11-02 13:49:16] piwik.WARNING: C:\matomo\core\Tracker\RequestSet.php(161): Warning - Trying to access array 
offset on value of type null - Matomo 4.11.0 - Please report this message in the Matomo 
forums: https://forum.matomo.org (please do a search first as it might have been reported already)
{
	"trace": [{
		"file": "C:\\matomo\\core\\Tracker\\RequestSet.php",
		"line": 161,
		"function": "errorHandler",
		"class": "Piwik\\ErrorHandler",
		"type": "::"
	}, {
		"file": "C:\\matomo\\plugins\\QueuedTracking\\Queue.php",
		"line": 94,
		"function": "restoreState",
		"class": "Piwik\\Tracker\\RequestSet",
		"type": "->"
	}, {
		"file": "C:\\matomo\\plugins\\QueuedTracking\\Queue\\Processor.php",
		"line": 112,
		"function": "getRequestSetsToProcess",
		"class": "Piwik\\Plugins\\QueuedTracking\\Queue",
		"type": "->"
	}, {
		"file": "C:\\matomo\\plugins\\QueuedTracking\\Commands\\Process.php",
		"line": 86,
		"function": "process",
		"class": "Piwik\\Plugins\\QueuedTracking\\Queue\\Processor",
		"type": "->"
	}, {
		"file": "C:\\matomo\\vendor\\symfony\\console\\Symfony\\Component\\Console\\Command\\Command.php",
		"line": 257,
		"function": "execute",
		"class": "Piwik\\Plugins\\QueuedTracking\\Commands\\Process",
		"type": "->"
	}, {
		"file": "C:\\matomo\\vendor\\symfony\\console\\Symfony\\Component\\Console\\Application.php",
		"line": 874,
		"function": "run",
		"class": "Symfony\\Component\\Console\\Command\\Command",
		"type": "->"
	}, {
		"file": "C:\\matomo\\vendor\\symfony\\console\\Symfony\\Component\\Console\\Application.php",
		"line": 195,
		"function": "doRunCommand",
		"class": "Symfony\\Component\\Console\\Application",
		"type": "->"
	}, {
		"function": "doRun",
		"class": "Symfony\\Component\\Console\\Application",
		"type": "->"
	}, {
		"file": "C:\\matomo\\core\\Console.php",
		"line": 135,
		"function": "call_user_func"
	}, {
		"file": "C:\\matomo\\core\\Access.php",
		"line": 670,
		"function": "Piwik\\{closure}",
		"class": "Piwik\\Console",
		"type": "->"
	}, {
		"file": "C:\\matomo\\core\\Console.php",
		"line": 136,
		"function": "doAsSuperUser",
		"class": "Piwik\\Access",
		"type": "::"
	}, {
		"file": "C:\\matomo\\core\\Console.php",
		"line": 87,
		"function": "doRunImpl",
		"class": "Piwik\\Console",
		"type": "->"
	}, {
		"file": "C:\\matomo\\vendor\\symfony\\console\\Symfony\\Component\\Console\\Application.php",
		"line": 126,
		"function": "doRun",
		"class": "Piwik\\Console",
		"type": "->"
	}, {
		"file": "C:\\matomo\\console",
		"line": 32,
		"function": "run",
		"class": "Symfony\\Component\\Console\\Application",
		"type": "->"
	}]
} {
	"class": "QueuedTracking",
	"request_id": 7812
}
```
The requests are gone from queue (almost everyone), but can´t see any statistics in Matomo during this test run.

**Comment by heurteph-ei** [comment]
I would say the request are not processed, because of the error...
Could you replay the 95 requests with QueueTracking disabled? Does it work? Or is the same error thrown?
In order to process the 2 last requests in queue, set 1 as nb of queue workers..., then run the queue processing then disable the queue processing...

**Comment by camlcase** [comment]
@heurteph-ei Got the same error, unfortunately.

**Comment by snake14** [comment]
Hi @camlcase . Could you please try adding the following below [line 91 of plugins/QueuedTracking/Queue.php](https://github.com/matomo-org/plugin-QueuedTracking/blob/4.x-dev/Queue.php#L91)?
```
if (empty($params)) {
    continue;
}
```

Your stack trace makes me think that you've somehow got invalid JSON and so `json_decode` simply returns null instead of the expected array.

**Comment by camlcase** [comment]
@snake14 I could. But wouldn't it just skip all, because all items seems already invalid in queue?

I wanna know why we're having these corrupt values in the first place. Do we need to change how we're sending tracking requests from our applications? If so, why and how? Or do we miss any fundamental stuff in the installation, configuration or setup?

We're planning right now for an upgrade, hoping it will solve this issue or maybe giving us some more information.

**Comment by snake14** [comment]
@camlcase I would hope that only 1 or 2 queued records are invalid and making that code change would allow the rest to process. That would also hopefully allow you to narrow down the exact record causing the error and then look at the contents to get a better idea of what went wrong.
If you have more than a couple invalid records then there's more to worry about and there could be some setup/configuration issue that needs to be tracked down. You could try waiting to see if the upgrade fixes the issue, but I don't know if it would if there's a record that's invalid JSON.

**Comment by djmetzle** [comment]
This does seem suspiciously similar to our problems from: https://github.com/matomo-org/plugin-QueuedTracking/issues/192

I do not believe that issue was adequately addressed.

Specifically, the two fixes included in the `4.12.x` patch releases:
https://github.com/matomo-org/matomo/pull/19855
https://github.com/matomo-org/device-detector/pull/7265

were not enough to get the queue processing. We were seeing thousands of requests pile up, as at the top of [@addison-grant's comment](https://github.com/matomo-org/plugin-QueuedTracking/issues/194#issuecomment-1299190865)

We additionally required a shim, as suggested in this comment:
 https://github.com/matomo-org/plugin-QueuedTracking/issues/192#issuecomment-1278240701

 > @djmetzle It definitely seems like you have some old queued records that are holding things up because they're in a slightly different format. What if you try something like this?
> 
> ```
> $varType = is_string($this->params['uadata']) ? 'json' : null;
> $clientHints = Common::getRequestVar('uadata', $default, $varType, $this->params);
> ```

On 4.12.0, manually including the two patches (from the 4.12.x patch releases) AND the `varType` ternary got us able to process and drain our queue.

I have a backlog issue for our internal infrastructure to build a `patch` into our matomo deployment, for just that `varType` fix. 

Please advise.

**Comment by snake14** [comment]
@djmetzle I thought that they might be related as well, but your issues seemed to start with the 4.12.0 upgrade and @camlcase is running version 4.11.0. Do you still need that adjustment that I suggested about the varType? I thought that it would only need to be temporary to get the backlog of old queue records created before the 4.12.0 upgrade cleared out. If there are new records being created that still need that vartType change, then I agree that the issue hasn't been fully addressed. Are you seeing any new errors in your logs?

**Comment by djmetzle** [comment]
@snake14 that must be a misunderstanding.

We started seeing queued requests fail to be processed, and get "stuck" pending in the queue, only after the 4.12 upgrade. Only 4.12 tracker requests were getting "stuck". Due to the way we deploy redis in our stack, no 4.11 queued tracker requests were being processed by 4.12. I can assure that the issue was not due to cross version compatibility. Something about processing the results from specifically 4.12 and the patch releases is broken. We need the `varType` fix to get our queue drained on 4.12, at all. 

The `varType` fix seems to fix type problems (or data corruption?) in the `ua` fields from device detector.

**Comment by djmetzle** [comment]
This comment has a dump of a tracker request that fails to be processed:
https://github.com/matomo-org/plugin-QueuedTracking/issues/192#issuecomment-1278275990

**Comment by snake14** [comment]
Sorry about the confusion @djmetzle . @sgiehl It sounds like an additional patch to 4.12 might be necessary unless you think there's a way to fix the issue in the QueuedTracking plugin. Please refer to the following comments for more information:
https://github.com/matomo-org/plugin-QueuedTracking/issues/194#issuecomment-1302459100
https://github.com/matomo-org/plugin-QueuedTracking/issues/194#issuecomment-1302693089

**Comment by camlcase** [comment]
@snake14 You had a good point there. I did some tests using your suggestion:

Queue when start processing requests:
`110 (110) request sets left in queue. 771.01K used memory (1.01M peak). 0 workers active.`

Debug info:
```
Starting to process request sets, this can take a while
[2022-11-04 10:26:12] piwik.DEBUG: Acquired lock for queue 0 [] {"class":"QueuedTracking","request_id":8652}

[2022-11-04 10:26:12] piwik.DEBUG: Acquired lock for queue 0 [] {"class":"QueuedTracking","request_id":8652}

[2022-11-04 10:26:12] piwik.DEBUG: Acquired lock for queue 0 [] {"class":"QueuedTracking","request_id":8652}


This worker finished queue processing with 0req/s (0 requests in 0.04 seconds)
```

No warnings but queue keeps increasing without any requests got processed.

**Comment by camlcase** [comment]
Upgrade to 4.12.3 done. I did the same tests again but still the same output: none of the queued requests got processed. The items in queue seems still broken. Values can´t be retrieved and decompressed.

**Comment by snake14** [comment]
Thank you for the updates @camlcase . Any recommendations @AltamashShaikh ?

**Comment by djmetzle** [comment]
Just checking, on `4.12.2` we're seeing the same typeerror from queuedtracking:
```
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] Uncaught exception: TypeError: json_decode(): Argument #1 ($json) must be of type string, array given in /var/www/html/core/Common.php:529
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] Stack trace:
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #0 /var/www/html/core/Common.php(529): json_decode(Array, true)
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #1 /var/www/html/core/Tracker/Request.php(643): Piwik\Common::getRequestVar('uadata', Array, 'json', Array)
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #2 /var/www/html/core/Tracker/VisitExcluded.php(193): Piwik\Tracker\Request->getClientHints()
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #3 /var/www/html/core/Tracker/VisitExcluded.php(67): Piwik\Tracker\VisitExcluded->isNonHumanBot()
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #4 /var/www/html/core/Tracker/Failures.php(55): Piwik\Tracker\VisitExcluded->isExcluded()
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #5 /var/www/html/core/Tracker/Visit.php(111): Piwik\Tracker\Failures->logFailure(1, Object(Piwik\Tracker\Request))
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #6 /var/www/html/core/Tracker/Visit.php(149): Piwik\Tracker\Visit->checkSiteExists(Object(Piwik\Tracker\Request))
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #7 /var/www/html/core/Tracker.php(172): Piwik\Tracker\Visit->handle()
ERROR Piwik\ExceptionHandler[2022-11-08 20:40:38 UTC] [41d0c] #8 /var/www/html/plugins/QueuedTracking/Queue/Processor/Handler.php(46): Piwik\Tracker->trackRequest(Object(Piwik\Tracker\Request))
```

and confirmed, this varType fix gets our queue processing again:
```php
$varType = is_string($this->params['uadata']) ? 'json' : null;
$clientHints = Common::getRequestVar('uadata', $default, $varType, $this->params);
```
from @snake14 in this comment: https://github.com/matomo-org/plugin-QueuedTracking/issues/192#issuecomment-1278240701

**Comment by sgiehl** [comment]
A change in core might fix this for sure. But we need to figure out why the uadata parameter already is any array when using queuetracking. This shouldn't be the case. Simply applying a patch in core might possibly hide the original problem.
The uadata parameter is provided as a json encoded string. Might be possible that the plugin decodes this, while it shouldn't.

**Comment by djmetzle** [comment]
@sgiehl good point. I hadnt been distinguishing between the core and plugin code, for the sake of a fix. Sorry about that.

**Comment by snake14** [comment]
@djmetzle Stefan's comment got me thinking and I came up with a fix in the plugin. It still isn't addressing the root cause of why the `uadata` parameter is already decoded in the fist place, but it should at least keep things processing without needing any additional core changes. Do you want to try rolling back the manual patch in `core/Tracker/Request.php` and try applying [the change from my new PR](https://github.com/matomo-org/plugin-QueuedTracking/pull/198/files#diff-851adce63cd615df56a1d397a89db035adfbdf2cf52dc5797199768230f6b957)? If that works, we can get my PR reviewed by one of my team members.

**Comment by djmetzle** [comment]
Awesome. I just applied the patch from https://github.com/matomo-org/plugin-QueuedTracking/pull/198 and it looks like the queue is processing now. 

I think that https://github.com/matomo-org/plugin-QueuedTracking/issues/192 was effectively addressed by #198 as well (that issue is already closed) Very nice! Thank you @snake14

**Comment by snake14** [comment]
@djmetzle That's great news. As I mentioned, it's still not addressing the root cause, but it's a step in the right direction.

@camlcase Do those changes fix queued tracking for you or is further investigation necessary?

**Comment by djmetzle** [comment]
Thanks guys!

**Comment by camlcase** [comment]
Sorry for late response. Been a busy bee here. Did a test using the suggested solution in #198 but got the same exception as before. The queued item can´t be decompressed into something readable.

`WARNING QueuedTracking[2022-11-15 13:39:49 UTC] [9964] C:\matomo\core\Tracker\RequestSet.php(161): Warning - Trying to access array offset on value of type null - Matomo 4.12.3 - Please report this message in the Matomo forums: https://forum.matomo.org (please do a search first as it might have been reported already) #0\core\Tracker\RequestSet.php(161),#1\plugins\QueuedTracking\Queue.php(98),#2\plugins\QueuedTracking\Queue\Processor.php(112),#3\plugins\QueuedTracking\Commands\Process.php(86),#4\vendor\symfony\console\Symfony\Component\Console\Command\Command.php(257),#5\vendor\symfony\console\Symfony\Component\Console\Application.php(874),#6\vendor\symfony\console\Symfony\Component\Console\Application.php(195),[internal function]: Symfony\Component\Console\Application->doRun(),#8\core\Console.php(135),#9\core\Access.php(670)`

**Comment by snake14** [comment]
@camlcase I'm sorry to hear that your issue hasn't been resolved. Could you please try adding the following just after [Queue line 94](https://github.com/matomo-org/plugin-QueuedTracking/blob/4.x-dev/Queue.php#L94)?
```
if ($params === null && json_last_error() !== JSON_ERROR_NONE) {
    $logger = StaticContainer::get('Psr\Log\LoggerInterface');
    $logger->error('Error decoding JSON. Error code: ' . json_last_error() . '. Error message: ' .json_last_error_msg() . '. Bad JSON string: ' . $value);
    continue;
}
```

You could also just use `throw new \Exception()` instead of logging, but then it will prevent any valid requests from processing. I'm curious how many invalid records you have and I'm also curious to see what the invalid JSON string looks like.

**Comment by shakalandy** [comment]
Having a similar issue with unprocessable Queue Entries as well over here:

```

[2023-03-22 09:27:49] piwik.ERROR: Uncaught exception: TypeError: json_decode(): Argument #1 ($json) must be of type string, array given in /var/www/stats.example.com/core/Common.php:529 Stack trace: #0 /var/www/stats.example.com/core/Common.php(529): json_decode() #1 /var/www/stats.example.com/core/Tracker/Request.php(448): Piwik\Common::getRequestVar() #2 /var/www/stats.example.com/core/Tracker/GoalManager.php(407): Piwik\Tracker\Request->getParam() #3 /var/www/stats.example.com/core/Tracker/GoalManager.php(376): Piwik\Tracker\GoalManager->getEcommerceItemsFromRequest() #4 /var/www/stats.example.com/core/Tracker/GoalManager.php(309): Piwik\Tracker\GoalManager->recordEcommerceGoal() #5 /var/www/stats.example.com/plugins/Goals/Tracker/GoalsRequestProcessor.php(108): Piwik\Tracker\GoalManager->recordGoals() #6 /var/www/stats.example.com/core/Tracker/Visit.php(223): Piwik\Plugins\Goals\Tracker\GoalsRequestProcessor->recordLogs() #7 /var/www/stats.example.com/core/Tracker.php(172): Piwik\Tracker\Visit->handle() #8 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor/Handler.php(46): Piwik\Tracker->trackRequest() #9 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor.php(194): Piwik\Plugins\QueuedTracking\Queue\Processor\Handler->process() #10 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor.php(143): Piwik\Plugins\QueuedTracking\Queue\Processor->processRequestSets() #11 /var/www/stats.example.com/plugins/QueuedTracking/Commands/Process.php(86): Piwik\Plugins\QueuedTracking\Queue\Processor->process() #12 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Command/Command.php(257): Piwik\Plugins\QueuedTracking\Commands\Process->execute() #13 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(874): Symfony\Component\Console\Command\Command->run() #14 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(195): Symfony\Component\Console\Application->doRunCommand() #15 /var/www/stats.example.com/core/Console.php(108): Symfony\Component\Console\Application->doRun() #16 [internal function]: Piwik\Console->originDoRun() #17 /var/www/stats.example.com/core/Console.php(147): call_user_func() #18 /var/www/stats.example.com/core/Access.php(670): Piwik\Console->Piwik\{closure}() #19 /var/www/stats.example.com/core/Console.php(148): Piwik\Access::doAsSuperUser() #20 /var/www/stats.example.com/core/Console.php(87): Piwik\Console->doRunImpl() #21 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(126): Piwik\Console->doRun() #22 /var/www/stats.example.com/console(32): Symfony\Component\Console\Application->run() #23 {main} {"exception":"[object] (TypeError(code: 0): json_decode(): Argument #1 ($json) must be of type string, array given at /var/www/stats.example.com/core/Common.php:529)","ignoreInScreenWriter":true} {"class":"Piwik\\ExceptionHandler","request_id":1161039}

Uncaught exception in /var/www/stats.example.com/core/Common.php line 529:
json_decode(): Argument #1 ($json) must be of type string, array given
#0 /var/www/stats.example.com/core/Common.php(529): json_decode()
#1 /var/www/stats.example.com/core/Tracker/Request.php(448): Piwik\Common::getRequestVar()
#2 /var/www/stats.example.com/core/Tracker/GoalManager.php(407): Piwik\Tracker\Request->getParam()
#3 /var/www/stats.example.com/core/Tracker/GoalManager.php(376): Piwik\Tracker\GoalManager->getEcommerceItemsFromRequest()
#4 /var/www/stats.example.com/core/Tracker/GoalManager.php(309): Piwik\Tracker\GoalManager->recordEcommerceGoal()
#5 /var/www/stats.example.com/plugins/Goals/Tracker/GoalsRequestProcessor.php(108): Piwik\Tracker\GoalManager->recordGoals()
#6 /var/www/stats.example.com/core/Tracker/Visit.php(223): Piwik\Plugins\Goals\Tracker\GoalsRequestProcessor->recordLogs()
#7 /var/www/stats.example.com/core/Tracker.php(172): Piwik\Tracker\Visit->handle()
#8 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor/Handler.php(46): Piwik\Tracker->trackRequest()
#9 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor.php(194): Piwik\Plugins\QueuedTracking\Queue\Processor\Handler->process()
#10 /var/www/stats.example.com/plugins/QueuedTracking/Queue/Processor.php(143): Piwik\Plugins\QueuedTracking\Queue\Processor->processRequestSets()
#11 /var/www/stats.example.com/plugins/QueuedTracking/Commands/Process.php(86): Piwik\Plugins\QueuedTracking\Queue\Processor->process()
#12 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Command/Command.php(257): Piwik\Plugins\QueuedTracking\Commands\Process->execute()
#13 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(874): Symfony\Component\Console\Command\Command->run()
#14 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(195): Symfony\Component\Console\Application->doRunCommand()
#15 /var/www/stats.example.com/core/Console.php(108): Symfony\Component\Console\Application->doRun()
#16 [internal function]: Piwik\Console->originDoRun()
#17 /var/www/stats.example.com/core/Console.php(147): call_user_func()
#18 /var/www/stats.example.com/core/Access.php(670): Piwik\Console->Piwik\{closure}()
#19 /var/www/stats.example.com/core/Console.php(148): Piwik\Access::doAsSuperUser()
#20 /var/www/stats.example.com/core/Console.php(87): Piwik\Console->doRunImpl()
#21 /var/www/stats.example.com/vendor/symfony/console/Symfony/Component/Console/Application.php(126): Piwik\Console->doRun()
#22 /var/www/stats.example.com/console(32): Symfony\Component\Console\Application->run()
#23 {main}

```

Matomo Version is the newest todate (4.14.1).

How can I delete the broken item from the queue?
How to find out, how to get the corresponding entry in redis to delete?

Seems to be error data inserted, as the patched mentioned here are already applied in the new release, as far as I've found out.

**Comment by shakalandy** [comment]
In case anyone else is stumbling upon this issue here.
The keys are logged in redis in one single key, default for the first queue is "trackingQueueV1".
You'll find all the values inside, and could delete the one with the lowest ID (usually 1), the queue will then proceed to process.

**Comment by snake14** [comment]
Hi @shakalandy . Thank you for providing more information. It looks like you just upgraded to the latest version of Matomo. What version were you running before that? Did you recently upgrade the QueuedTracking plugin as well? Also, could you please provide the versions that you are running for the following?
PHP:
Redis:
QueuedTracking plugin:

**Comment by shakalandy** [comment]
Sure, happy to help.

> What version were you running before that? 

The latest 4.13.3 as clean install some days ago, before the 4.14.0 release

> Did you recently upgrade the QueuedTracking plugin as well?

it was freshly installed, with 4.13.3 above.  

> Also, could you please provide the versions that you are running for the following?

PHP: 8.1.17 (8.1.17-1+ubuntu20.04.1+deb.sury.org+1 to be more specific)
Redis: 5.0.7-2ubuntu0.1
QueuedTracking plugin: v4.0.6

I think my client tried to insert e-commerce tracking data via the API.
So as a guess I think the JSON validation is now better for normal requests with all these fixes, but maybe they are missing on the e-commerce tracking as well?

**Comment by snake14** [comment]
Thank you again @shakalandy . That is very helpful. The fact that it's a recent install and the latest version of the plugin is quite interesting. @sgiehl Does this provide any new insights on where the back tracking records are coming from?

**Comment by sgiehl** [comment]
@shakalandy might it be possible that you passed the ecommerce items (`ec_items`) directly as an array parameter in the tracking url? That parameter is expected to be passed as json_encoded string?

**Comment by pwe-sg** [comment]
We have the same Issue on our end. We just set up a new Matomo server and the queue isn't getting processed.

Matomo Version: 4.14.1
Queued-Tracking Version: v4.0.6
PHP Version: 7.4.33
Redis Version: 5:6.0.16-1+deb11u2

When we disable the queue the requests are getting processed by Matomo just fine. If the queue is activated the requests are getting stored in the queue but when I manually want to process them I get the message:

> This worker finished queue processing with 0req/s (0 requests in 0.00 seconds)

**Comment by mind63** [comment]
@sgiehl It was me who send the data and yes, I passed ec_items as array while building a request in PHP. Of course it was not the cleanest way and the array was a mistake, but the consequence is pretty heavy. I crashed the complete Matomo process with just one unclear request. We had to delete the record manually from the Redis-Queue (it was the first/oldest entry).

I think Matomo should just skip incorrect entries.

**Comment by sgiehl** [comment]
@mind63 I agree that matomo should simply ignore request that it can't handle instead of crashing. I just wanted to be sure that Matomo did not generate that request, as that would have been a way more critical then.
Not sure if we can catch thus errors in the QueueTracking plugin, maybe log the error and remove the entries from the queue nevertheless, so the queue is at least able to process.

**Comment by snake14** [comment]
@sgiehl I could try ensuring that `ec_items` is a JSON string, [similar to what we did for uadata](https://github.com/matomo-org/plugin-QueuedTracking/commit/6e08c5bb23b5bdccbf3eb0c0d7977cb41d1aae02). However, trying to fix bad inputs from the QueuedPlugin side of things feels a little like playing whack-a-mole. Isn't it the responsibility of an API to validate its inputs? I would have expected the API to catch invalid input and prevent the invalid data from being processed or inserted into the database.

Edit: Thinking about it more, I still think that the API should validate its inputs but I also think you have a good point about one error stopping all requests from being processed. I can create a new issue to evaluate how the plugin currently handles errors and how we can improve it to allow skipping requests. There are some concerns with that though. How do we adequately surface that to the customer? How do we prevent too many bad requests from building up in the database and slowing things down? If a request is skipped and later fixed, would that throw off the order of requests, since the plugin wouldn't have processed them in chronological order?

Do you have any thoughts to add @AltamashShaikh ?

**Comment by sgiehl** [comment]
@snake14 I didn't mean to check if the parameters are sent with correct types. I agree that the plugin shouldn't do an type of validation. 
I'm not much into the code of this plugin, but looking at the stack traces of the error reported above it looks like there is some sort of queue processor, which handles every requests one by one. 
If one of the requests currently throws an php error the whole queue processing is aborted. What I would suggest is to change the try/catch around this processing here:
https://github.com/matomo-org/plugin-QueuedTracking/blob/8855ec5cbace81339aae0fcc8d8acfd4dc5b4be8/Queue/Processor/Handler.php#L43-L56
Adding an `catch (\Error $e)` would allow to skip requests that throw a catchable php error.

**Comment by snake14** [comment]
Sounds good @sgiehl . I was looking at the same try/catch. I would just create a quick PR, but I think we need to think things through to make sure that we're not introducing regressions. I'll see if I can move this issue up in priority.

**Comment by sgiehl** [comment]
I guess everything is better than a fatal in processing the queue. Logging a warning might be helpful to identify possible problems in Matomo though.
I'll nevertheless prepare an improvement around the json parameter handling for core.

**Comment by pwe-sg** [comment]
Seems like our queue was working just fine but with the default settings it just didn't get triggered. I lowered it to 2 worker which should start after 3 requests and it's working now.

But I cannot find any documentation how to use the queuedTracking:process. It seems to do nothing. This was the reason that I didn't lower the workers because it didn't even seem to get processed manually.

**Comment by snake14** [comment]
Hi @pwe-sg I'm glad that you were able to get your queue processing. For more documentation about using QueuedTracking, you could refer to the [related FAQ article](https://matomo.org/faq/on-premise/how-to-configure-matomo-to-handle-unexpected-peak-in-traffic/) or the [plugin marketplace](https://plugins.matomo.org/QueuedTracking#faq).

**Comment by atom-box** [comment]
Possibly related? #211 

> "**It's like Queued Tracking has a dangling request that cant be processed at the front of the queue** that holds up the rest of the queue.
> I solved it for now by opening up more queues, more workers.
> But queue 0 is still stuck.
> And eventually it will fill up memory and crash."

**Comment by snake14** [comment]
The fix for this issue should be included in the next release of this plugin.
